packages/slate-react/src/components/editor.tsx

```diff
@@ -0,0 +1,488 @@
+  // to the real event sadly. (2019/11/01)
+  // https://github.com/facebook/react/issues/11211
+  const onDOMBeforeInput = useCallback(
+    (
+      event: Event & {
+        data: string | null
+        dataTransfer: DataTransfer | null
+        getTargetRanges(): DOMStaticRange[]
+        inputType: string
+        isComposing: boolean
+      }
+    ) => {
+      if (!readOnly && hasEditableTarget(editor, event.target)) {
+        console.log('beforeinput', event)
+        const { inputType } = event
+
+        // These two types occur while a user is composing text and can't be
+        // cancelled. Let them through and wait for the composition to end.
+        if (
+          inputType === 'insertCompositionText' ||
+          inputType === 'deleteCompositionText'
+        ) {
+          return
+        }
+
+        const [targetRange] = event.getTargetRanges()
+
+        if (targetRange && inputType !== 'deleteContent') {
+          const range = editor.toSlateRange(targetRange)
+          editor.select(range)
+        }
+
+        event.preventDefault()
+        editor.onBeforeInput(event)
+      }
+    },
+    []
+  )
+
+  // On native `selectionchange` event, trigger the `onSelect` handler. This is
+  // needed to account for React's `onSelect` being non-standard and not firing
+  // until after a selection has been released. This causes issues in situations
+  // where another change happens while a selection is being dragged.
+  const onDOMSelectionChange = useCallback(
+    debounce(() => {
+      console.log('selectionchange')
+      if (!readOnly && !state.isComposing && !state.isUpdatingSelection) {
+        const el = editor.toDomNode(value)
+        const { activeElement } = window.document
+
+        if (activeElement === el) {
+          const domSelection = window.getSelection()
+          state.latestElement = activeElement
+
+          if (domSelection) {
+            const range = editor.toSlateRange(domSelection)
+            console.log('select', domSelection, range)
+            editor.select(range)
+            IS_FOCUSED.set(editor, true)
+          }
+        } else {
+          IS_FOCUSED.delete(editor)
+        }
+      }
+    }, 100),
+    []
+  )
+
+  return (
+    <EditorContext.Provider value={editor}>
+      <ReadOnlyContext.Provider value={readOnly}>
+        <FocusedContext.Provider value={editor.isFocused()}>
+          <div
+            // COMPAT: The Grammarly Chrome extension works by changing the DOM
+            // out from under `contenteditable` elements, which leads to weird
+            // behaviors so we have to disable it like editor. (2017/04/24)
+            data-gramm={false}
+            role={readOnly ? undefined : props.role || 'textbox'}
+            {...attributes}
+            data-slate-editor
+            data-slate-node="value"
+            contentEditable={readOnly ? undefined : true}
+            suppressContentEditableWarning
+            ref={ref}
```

Comment:
I was thinking, and maybe this can be a feature to worry about as a future item, that you could wrap the editor in [`forwardRef`](https://reactjs.org/docs/forwarding-refs.html). Then you can combine that ref with the ref used internally to allow users to pass in their own ref to attach to the editor.

Reply:
Sounds good to me.